woocommerce/google-listings-and-ads
# Filters for manually mapping product IDs

### Changes proposed in this Pull Request:

This PR adds two filters that permit manual mapping of the WooCommerce product ID to the Merchant Center product ID (`woocommerce_gla_get_google_product_offer_id`),and vice versa (`woocommerce_gla_get_wc_product_id`). The functionality should allow merchants with existing catalogs on Merchant Center to specify which WooCommerce products correspond to which existing MC products, which will likely not have IDs that follow the GLA structure (`gla_1234` for product 1234).

**Note:** I've been poring over the code to see if there are any places where `product_id` or `mc_id` or `slug` are used in a context of WooCommerce product ID <=> Merchant Center product ID. Nothing has stuck out to me, but it would be nice to have another set of eyes, especially regarding reporting and Ads.

### Detailed test instructions:
**WC to MC**
1. Fast test: 
    1. Add a quick filter to map Woo IDs to custom MC IDs (the actual product IDs don't matter):
	```
	add_filter( 'woocommerce_gla_get_google_product_offer_id', function( $mc_id, $woo_id ) {
		$woo_to_mc = [
			10 => 'test_id_1',
			20 => 'test_id_2',
			30 => 'gla_5',
		];
		if( is_numeric( $woo_id ) && isset( $woo_to_mc[$woo_id] ) ) {
			$mc_id = $woo_to_mc[$woo_id];
		}
		return $mc_id;
	}, 10, 2 );
	```

    2. Test some of the returned MC IDs using `WCProductAdapter::get_google_product_offer_id`:
	```
	add_action( 'init', function(){
		// Enable WP_DEBUG and WP_DEBUG_LOG
		foreach( [ 10, 20, 30, 40 ] as $id ) {
			$mapped_id = Automattic\WooCommerce\GoogleListingsAndAds\Product\WCProductAdapter::get_google_product_offer_id( 'gla', $id );
			error_log( "$id mapped to $mapped_id"  );
		}
		// Log should read:
		// 10 mapped to test_id_1
		// 20 mapped to test_id_2
		// 30 mapped to gla_5
		// 40 mapped to gla_40
	});
	```
2. Or, longer test:
    1. Add the filter listed in _1.i_, but using real product IDs
    2. Connect to a Merchant Center account and wait for product sync or force sync from the Test Connection page.
    3. Confirm that the mapped products have been created in Merchant Center with their custom IDs.

**MC to WC**
1. Add a filter to map custom MC IDs to Woo IDs (the product IDs don't matter):
	```
	add_filter( 'woocommerce_gla_get_wc_product_id', function( $mc_product_id, $prefix ) {
		$mc_to_woo = array_flip( [
			10 => 'test_id_1',
			20 => 'test_id_2',
			30 => 'gla_5',
		] );
		if ( isset( $mc_to_woo[$mc_product_id] ) ) {
			$mc_product_id = $prefix . $mc_to_woo[$mc_product_id];
		}
		return $mc_product_id;
	}, 10, 2 );
	```
    2. Test some of the returned Woo IDs using `ProductHelper::get_wc_product_id`:
    ```
	add_action( 'init', function() {
		// Enable WP_DEBUG and WP_DEBUG_LOG
		$product_helper = woogle_get_container()->get( Automattic\WooCommerce\GoogleListingsAndAds\Product\ProductHelper::class );
		foreach ( [ 'test_id_1', 'test_id_2', 'gla_5', 'gla_40' ]  as $id ) {
			error_log( "$id mapped to " . $product_helper->get_wc_product_id( $id ) );
		}
		// Log should read:
		// test_id_1 mapped to 10
		// test_id_2 mapped to 20
		// gla_5 mapped to 30
		// gla_40 mapped to 40
	});
    ```


### Changelog entry

> Add - Filters for manually mapping product IDs.
